drivers/pwm/pwm_stm32.c

```diff
@@ -16,26 +15,108 @@
 
 #include <drivers/clock_control/stm32_clock_control.h>
 
-#include "pwm_stm32.h"
-
 #include <logging/log.h>
-LOG_MODULE_REGISTER(pwm_stm32);
+LOG_MODULE_REGISTER(pwm_stm32, CONFIG_PWM_LOG_LEVEL);
+
+/** PWM data. */
+struct pwm_stm32_data {
+	/** Timer clock (Hz). */
+	u32_t tim_clk;
+};
+
+/** PWM configuration. */
+struct pwm_stm32_config {
+	/** Timer instance. */
+	TIM_TypeDef *timer;
+	/** Prescaler. */
+	u32_t prescaler;
+	/** Clock configuration. */
+	struct stm32_pclken pclken;
+};
+
+static inline struct pwm_stm32_data *to_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct pwm_stm32_config *to_config(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+/**
+ * Obtain LL channel given numeric channel.
+ *
+ * @param channel    Timer channel.
+ * @param ll_channel LL Timer channel.
+ *
+ * @return 0 if valid, -EINVAL if not.
+ */
+static int get_channel(u32_t channel, u32_t *ll_channel)
+{
+	if (channel == 1u) {
+		*ll_channel = LL_TIM_CHANNEL_CH1;
+	} else if (channel == 2u) {
+		*ll_channel = LL_TIM_CHANNEL_CH2;
+	} else if (channel == 3u) {
+		*ll_channel = LL_TIM_CHANNEL_CH3;
+	} else if (channel == 4u) {
+		*ll_channel = LL_TIM_CHANNEL_CH4;
+#if !defined(CONFIG_SOC_SERIES_STM32L0X) &&                                    \
+	!defined(CONFIG_SOC_SERIES_STM32L1X) &&                                \
+	!defined(CONFIG_SOC_SERIES_STM32F0X) &&                                \
+	!defined(CONFIG_SOC_SERIES_STM32F1X) &&                                \
+	!defined(CONFIG_SOC_SERIES_STM32F2X)
+	} else if (channel == 5u) {
+		*ll_channel = LL_TIM_CHANNEL_CH5;
+	} else if (channel == 6u) {
+		*ll_channel = LL_TIM_CHANNEL_CH6;
+#endif
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
 
-/* convenience defines */
-#define DEV_CFG(dev)							\
-	((const struct pwm_stm32_config * const)(dev)->config->config_info)
-#define DEV_DATA(dev)							\
-	((struct pwm_stm32_data * const)(dev)->driver_data)
-#define PWM_STRUCT(dev)					\
-	((TIM_TypeDef *)(DEV_CFG(dev))->pwm_base)
+/**
+ * Obtain LL polarity from PWM flags.
+ *
+ * @param flags PWM flags.
+ *
+ * @return LL polarity.
+ */
+static u32_t get_polarity(pwm_flags_t flags)
+{
+	if (flags & PWM_POLARITY_NORMAL) {
+		return LL_TIM_OCPOLARITY_HIGH;
+	}
```

Comment:
When you add support for PWM flags here, you should also add support for flags to the PWM DTS binding (see https://github.com/zephyrproject-rtos/zephyr/blob/ccb5a01fc18ca5e8879cfa21adbb77b9740a6ac5/dts/bindings/pwm/nxp%2Ckinetis-ftm.yaml#L24 for example).

Reply:
Thanks @henrikbrixandersen! I have pushed a new commit with updated DT bindings as well as affected boards.